test/Microsoft.ML.Functional.Tests/ONNX.cs

```diff
@@ -0,0 +1,195 @@
+        /// </summary>
+        [OnnxFactAttribute]
+        public void SaveOnnxModelLoadAndScoreSDCA()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            // Get the dataset.
+            var data = mlContext.Data.LoadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Transforms.Normalize("Features"))
+                .AppendCacheCheckpoint(mlContext)
+                .Append(mlContext.Regression.Trainers.Sdca(
+                    new SdcaRegressionTrainer.Options { NumberOfThreads = 1, MaximumNumberOfIterations = 10 }));
+
+            // Fit the pipeline.
+            var model = pipeline.Fit(data);
+
+            // Serialize the pipeline to a file.
+            var modelFileName = "SaveOnnxLoadAndScoreSdcaModel.onnx";
+            var modelPath = DeleteOutputPath(modelFileName);
+            using (var file = File.Create(modelPath))
+                mlContext.Model.ConvertToOnnx(model, data, file);
+
+            // Load the model as a transform.
+            var onnxEstimator = mlContext.Transforms.ApplyOnnxModel(modelPath);
+            var onnxModel = onnxEstimator.Fit(data);
+
+            // TODO #2980: ONNX outputs don't match the outputs of the model, so we must hand-correct this for now.
+            // TODO #2981: ONNX models cannot be fit as part of a pipeline, so we must use a workaround like this.
+            var onnxWorkaroundPipeline = onnxModel.Append(
+                mlContext.Transforms.CopyColumns("Score", "Score0").Fit(onnxModel.Transform(data)));
+
+            // Create prediction engine and test predictions.
+            var originalPredictionEngine = model.CreatePredictionEngine<HousingRegression, ScoreColumn>(mlContext);
+            // TODO #2982: ONNX produces vector types and not the original output type.
+            var onnxPredictionEngine = onnxWorkaroundPipeline.CreatePredictionEngine<HousingRegression, OnnxScoreColumn>(mlContext);
+
+            // Take a handful of examples out of the dataset and compute predictions.
+            var dataEnumerator = mlContext.Data.CreateEnumerable<HousingRegression>(mlContext.Data.TakeRows(data, 5), false);
+            foreach (var row in dataEnumerator)
+            {
+                var originalPrediction = originalPredictionEngine.Predict(row);
+                var onnxPrediction = onnxPredictionEngine.Predict(row);
+                // Check that the predictions are identical.
+                Assert.Equal(originalPrediction.Score, onnxPrediction.Score[0], precision: 4); // Note the low-precision equality!
+            }
+        }
+
+        private class ScoreColumn
+        {
+            public float Score { get; set; }
+        }
+
+        private class OnnxScoreColumn
```

Comment:
What's the difference between `OnnxScoreColumn` and `ClusteringScoreColumn`? Without specifying the dimension of `Score` field, I am not sure the code is safe. Maybe we can do
```csharp
[VectorType(dimension)]
public float[] Score { get; set; }
```

Reply:
The `ScoreColumn` class will be replaced with a general one being added by the `ModelFiles` PR.

The `OnnxScoreColumn` is there to make it explicit that we are working around a ?bug? in our ONNX implementation.

The `ClusteringScoreColumn` I added specifically for clustering.

Having wrote that out, we can delete the last two and make a `ScoreArrayColumn` class in the general helper class files (in `ModelFiles`, but I'll pull it into this PR instead and rebase that one).

On the topic of `VectorType`, we don't need to specify the dimension. Specifying a dimension just guarantees that the vector will be the same length for each row. That has the downside of making the classes non-reusable, so for helper classes in tests, we usually don't specify this attribute.

---
In reply to: [266636611](https://github.com/dotnet/machinelearning/pull/2984#discussion_r266636611) [](ancestors = 266636611)